src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs

```diff
@@ -140,115 +140,112 @@ public class HubEndPoint<THub, TClient> : EndPoint, IInvocationBinder where THub
+                await Execute(connection, invocationMessage);
             }
             catch (Exception ex)
             {
                 // Set the exception on the task completion source
-                tcs.TrySetException(ex);
+                dispatcherCompletion.TrySetException(ex);
 
                 // Cancel reading operation
-                cts.Cancel();
+                dispatcherCancellation.Cancel();
             }
         }
 
-        private async Task Execute(Connection connection, IInvocationAdapter invocationAdapter, InvocationDescriptor invocationDescriptor)
+        private async Task Execute(Connection connection, InvocationMessage invocationMessage)
         {
-            InvocationResultDescriptor result;
             HubMethodDescriptor descriptor;
-            if (_methods.TryGetValue(invocationDescriptor.Method, out descriptor))
+            if (!_methods.TryGetValue(invocationMessage.Target, out descriptor))
             {
-                result = await Invoke(descriptor, connection, invocationDescriptor);
+                // Send an error to the client. Then let the normal completion process occur
+                _logger.LogError("Unknown hub method '{method}'", invocationMessage.Target);
+                await SendMessageAsync(connection, ResultMessage.WithError(invocationMessage.InvocationId, $"Unknown hub method '{invocationMessage.Target}'"));
             }
             else
             {
-                // If there's no method then return a failed response for this request
-                result = new InvocationResultDescriptor
-                {
-                    Id = invocationDescriptor.Id,
-                    Error = $"Unknown hub method '{invocationDescriptor.Method}'"
-                };
-
-                _logger.LogError("Unknown hub method '{method}'", invocationDescriptor.Method);
+                var result = await Invoke(descriptor, connection, invocationMessage);
+                await SendMessageAsync(connection, result);
             }
 
-            // TODO: Pool memory
-            var outStream = new MemoryStream();
-            await invocationAdapter.WriteMessageAsync(result, outStream);
+            // Send the completion message
+            await SendMessageAsync(connection, new CompletionMessage(invocationMessage.InvocationId));
+        }
+
+        private async Task SendMessageAsync(Connection connection, HubMessage hubMessage)
+        {
+            // Encode the message
+            var payload = _protocol.WriteMessage(hubMessage);
 
-            var outMessage = new Message(outStream.ToArray(), MessageType.Text, endOfMessage: true);
+            // Create a message and put it on the outbound channel
+            var message = new Message(payload, _protocol.MessageType, endOfMessage: true);
 
-            while (await connection.Transport.Output.WaitToWriteAsync())
+            while (!connection.Transport.Output.TryWrite(message))
```

Comment:
Doing it in a single place is pretty confusing, revert after do it later, everywhere. Though I hate how it looks.